frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -1495,6 +1510,27 @@ impl<T: Config> Pallet<T> {
 		Ok(ReadySolution { supports, compute, score })
 	}
 
+	/// Should be used instead of the getter of `Snapshot` in memory-bounded code paths.
+	fn read_snapshot_with_preallocate() -> Result<RoundSnapshot<T>, Error<T>> {
+		use codec::MaxEncodedLen;
+		let snap = Self::snapshot_metadata().ok_or(Error::<T>::MissingSnapshot)?;
+		let voters_size = snap.voters as usize * <VoterOf<T>>::max_encoded_len();
+		let targets_size = snap.targets as usize * T::AccountId::max_encoded_len();
+
+		// we want to decode two vecs, which need two u32s at most for their size.
+		let initial_capacity = voters_size + targets_size + 4 + 4;
+		let mut buffer = Vec::<u8>::with_capacity(initial_capacity);
+
+		// fill this whole buffer, and decode into it.
+		buffer.resize(buffer.capacity(), 0);
+		let _leftover = sp_io::storage::read(&<Snapshot<T>>::hashed_key(), &mut buffer, 0)
+			.ok_or(Error::<T>::MissingSnapshot)?;
+
+		// buffer should have not re-allocated
+		debug_assert!(buffer.capacity() == initial_capacity);
+		<RoundSnapshot<T> as codec::Decode>::decode(&mut &*buffer).map_err(|_| Error::<T>::Codec)
+	}
```

Comment:
Hmm... so you manually calculate how much the scale-encoded value is supposed to take, preallocate a buffer, fetch it into that buffer and then decode it. Can you explain why is this necessary? How is this different than calling `Snapshot::get()`?

AFAIK this should already be done on the FFI boundary automatically, unless I'm missing something here? (And if I'm missing something then it'd be better to fix this on the FFI boundary so that it's fixed everywhere instead of manually fixing in on a case-by-case basis like here.)

Reply:
Yeah, the FFI should also allocate `buffer` only to the actual size it has in the storage. Meaning it should even take less memory than you are calculating here.

Reply:
I can provide some proof using the log that I added in the allocator, but I am pretty sure I consume less memory this way than calling `Snapshot::get()`.

Reply:
> I can provide some proof using the log that I added in the allocator, but I am pretty sure I consume less memory this way than calling `Snapshot::get()`.

This is how `get` works right now AFAIK:

1. The host function to grab the value from storage gets triggered.
2. The host reads the value from the storage into a host-side `Vec`.
3. The host calls into the runtime's allocator and directly allocates a new chunk of memory accessible *within* the runtime.
4. The host copies the data from the host-side `Vec` into the chunk of memory allocated at step (3).
5. The host packs the pointer to that chunk of runtime memory along with its size into a single `u64` value and returns that from the host function. (This is zero-cost as primitive types like `u64` are natively supported by the FFI.)
6. The runtime grabs that `u64` value, unpacks it into a (ptr, size) pair and does a `Vec::from_raw_parts` to create a `Vec<u8>`.
7. The runtime calls `Decode::decode` on that `Vec<u8>`.

So as you can see this shouldn't consume any extra memory since the host essentially directly allocates the runtime-side `Vec` into which it copies the data. It always allocates only the exact amount of bytes necessary, and directly copies the data without any extra resizes.

So either this code doesn't actually do anything different, *or* it does and the FFI layer doesn't work exactly as we think it does (in which case we should fix it).

So yeah, it'd be nice if you could double check that this actually does anything different.

Reply:
@koute your description reads exactly on how the implementation is/should be.